toolbox/metrics/main.go

```diff
@@ -0,0 +1,352 @@
+	"log"
+	"math"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"golang.org/x/net/context"
+)
+
+const (
+	fetchInterval = 5
+)
+
+var (
+	webAddress = flag.String("listen_port", ":9103", "port on which to expose metrics and web interface.")
+	jenkinsURL = flag.String("jenkins_url", "https://testing.istio.io", "URL to Jenkins API")
+	gcsBucket  = flag.String("bucket", "istio-code-coverage", "gcs bucket")
+
+	succeededBuilds = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: "succeeded_build_durations_seconds",
+			Help: "Succeeded build durations seconds.",
+		},
+		[]string{"build_job", "repo"},
+	)
+
+	failedBuilds = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: "failed_build_durations_seconds",
+			Help: "Failed build durations seconds.",
+		},
+		[]string{"build_job", "repo"},
+	)
+
+	succeededBuildMax = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "succeeded_build_durations_max_seconds",
+			Help: "Succeeded max build durations seconds.",
+		},
+		[]string{"build_job", "repo"},
+	)
+
+	succeededBuildMin = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "succeeded_build_durations_min_seconds",
+			Help: "Succeeded min build durations seconds.",
+		},
+		[]string{"build_job", "repo"},
+	)
+
+	codeCoverage = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "package_code_coverage",
+			Help: "Package code coverage.",
+		},
+		[]string{"package", "repo"},
+	)
+
+	jobRegister = map[string][]string{
+		"pilot":       {"presubmit", "postsubmit", "e2e-smoketest", "e2e-suite"},
+		"mixer":       {"presubmit", "postsubmit", "e2e-smoketest", "e2e-suite"},
+		"proxy":       {"presubmit", "postsubmit"},
+		"auth":        {"presubmit", "postsubmit"},
+		"mixerclient": {"presubmit"},
+		"istio":       {"presubmit"},
+		"test-infra":  {"presubmit"},
+	}
+
+	codeCovRegister = map[string]bool{
+		"pilot": true,
+		"mixer": true,
+		"auth":  true,
+	}
+
+	codeCovTrackJob = "presubmit"
+
+	gcsClient *storage.Client
+
+	jk = &jkConst{
+		building:      "building",
+		result:        "result",
+		duration:      "duration",
+		number:        "number",
+		resultFailure: "FAILURE",
+		resultSUCCESS: "SUCCESS",
+	}
+)
+
+type jkConst struct {
+	building      string
+	result        string
+	duration      string
+	number        string
+	resultSUCCESS string
+	resultFailure string
+}
+
+type repo struct {
+	name string
+	jobs []*job
+}
+
+type job struct {
+	repoName    string
+	jobName     string
+	lastBuildID int
+	client      *http.Client
+}
+
+func newJob(repoName, jobName string) *job {
+	return &job{
+		repoName:    repoName,
+		jobName:     jobName,
+		lastBuildID: -1,
+		client:      &http.Client{},
+	}
+}
+
+func newRepo(name string) *repo {
+	r := &repo{
+		name: name,
+		jobs: make([]*job, 0),
+	}
+	for n := range jobRegister[name] {
```

Comment:
yes, we get a nil if the key doesn't exist, and when n == nil, it won't go into the loop